### DISCUSSION TITLE
How to use Nuxt with newrelic node agent

### DISCUSSION THREAD

**Discussion by theomjones** [nuxt/nuxt#9303] [Questions (Nuxt 2)]
### What problem does this feature solve?

Newrelic's node agent has a requirement that it's the first line of your main module in order to properly bootstrap. https://docs.newrelic.com/docs/agents/nodejs-agent/installation-configuration/install-nodejs-agent#installing

Currently there's no way to do this in Nuxt's setup.

There may also be other use-cases outside of analytics modules, but I can't think of any right now.

### What does the proposed changes look like?

Support some way of requiring modules as the first line of the main js entrypoint, perhaps from a hook or something that will be injected in the build step? I would provide code examples but I'm not that familiar with Nuxt internals yet.

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c9646">#c9646</a>)</em></sub></div>

**Comment by pi0** [comment]
Hi. I suggest 3 approaches to try:

## 1. Config

We can try to add `import 'newrelic'` to the top of `nuxt.config.js`? This should partially work as this is almost the topmost entry-point. 

## 2. CLI Wrapper

We may create a custom cli wrapper in the root of the project:

`~/nuxt.js`:

```js
require('newrelic')
require('nuxt/bin/nuxt')
````

And then inside `package.json`:

```diff
scripts: {
-- "start": "nuxt start",
++ "start": "node ./nuxt start",
}
```

## 3. Module (best)

If newrelic is similar to sentry, this is not **required** to be the first import, we can do this via a nuxt module and it globally should register itself to the entire process:

`modules/newrelic.js`

```js
export default async function() {
  await import('newrelic')
}
```

`nuxt.config.js`

```js
export default {
    modules: [
        '~/modules/newrelic'
    ]
}
```

**Comment by kezek** [comment]
I'm interested in this as well, did you attempt anything @theomjones ?

Thanks

**Comment by theomjones** [comment]
Hi @kezek, yeah we went with the CLI wrapper as new relic has to be the first required module for full metrics.

**Comment by kezek** [comment]
Thanks for the feedback. Do you also use PM2 by any chance ?

**Comment by theomjones** [comment]
Yeah we do, we went by these docs: https://pm2.keymetrics.io/docs/usage/docker-pm2-nodejs/

**Comment by nachogarcia** [comment]
I used the CLI wrapper and it works properly.
Sadly, it's not integrating with the router and all the transactions are shown as /GET
I was thinking of [binding the name with the router (in a middleware maybe)](https://docs.newrelic.com/docs/agents/nodejs-agent/api-guides/nodejs-agent-api#request-api) but dind't have time to get my hands into it.

Has anyone faced that issue?

**Comment by nachogarcia** [comment]
After some hassle, I managed to set it up.

I added newrelic as an external webpack dependency in nuxt.config.js
```javascript
    extend (config, ctx) {
      config.externals = [
        { newrelic: 'newrelic' }
      ]
    }
```

And then in nuxtServerInit I set the transaction name with code split so the bundle doesn't break in client side

```javascript
async nuxtServerInit ({ dispatch, commit }, context) {
    const newrelic = await import('newrelic')
    newrelic.setTransactionName(context.route.matched[0].path)
}
```

**Comment by mchagas90** [comment]
@nachogarcia can you see the time cosuming at the external requests on newrelic site? I have the configuration just like yours, but can't see this info.

**Comment by nachogarcia** [comment]
Yeah, they show. Although for some reason I can only see the details of the calls to my PIM, not to my CMS (probably something related to the infra). It looks like this:
![image](https://user-images.githubusercontent.com/9975630/81649244-4c707180-9430-11ea-8587-6d792e3e5881.png)

**Comment by ghost** [comment]
@nachogarcia Hi, we tried your approach with implementing the two rows in 'nuxtServerInit' function to see the full path of Browser tracing names:
const newrelic = await import ('newrelic')
newrelic.SetTransactionName(context.route.matched[0].path)

We found out in our CI/CD process build, the script was broken every time when IIS physical path set to folder when we running the 'npm run build' script.

When we removed the two mentioned rows everything was running as well.

The error we get is:
ERROR in ./node_modules/newrelic/lib/system-info.js
Module not found: Error: Can't resolve 'child_process' in 'node_modules\newrelic\lib'

Can you assists please?
Thank you,
SP

**Comment by nachogarcia** [comment]
No idea. Did you set the external webpack dependency?

Also, I improved this a bit. There were some routes without match so now the line looks like
`newrelic.setTransactionName(context.route.matched[0] ? context.route.matched[0].path : context.route.name)`

And the GETs that didn't get logged were the calls to my healthcheck serverMiddleware so I added the env variable `NEW_RELIC_IGNORING_RULES: '^/health'`

**Comment by ghost** [comment]
1. We did set the external webpack dependency as you suggested. I would like to add that our NuxtJS app is running as a Windows service. We can't find a reason why the build fails to import newrelic in the nuxtServerInit. Do you have any other thoughts on that?

2. Thanks you for sharing your improvements. We'll try it once the above issue will be resolved.

**Comment by nachogarcia** [comment]
Doing a quick search it seems like it's a rather common issue with other libraries. People fix it with

```
module.exports.node = {
  child_process: 'empty'
}
```

Maybe you can go through that route, or you can open an issue in https://github.com/newrelic/node-newrelic

**Comment by ghost** [comment]
We tried child_process: 'false' and it didn't worked so I'm not sure if 'empty' will make a difference but I'll give it a shot.

Much appreciated.

**Comment by claudiolcastro** [comment]
@nachogarcia Hey! I've configured my newrelic as same as you did and its working fine, but on the client interactions, Im having this error:

`logger.js?899e:6 Uncaught (in promise) TypeError: newrelic.noticeError is not a function`

Do you have any idea of wahts possibly going on? thanks ;)

**Comment by nachogarcia** [comment]
We're using it on nuxtServerInit with the dynamic import for using server side only :/

**Comment by claudiolcastro** [comment]
> We're using it on nuxtServerInit with the dynamic import for using server side only :/

humm I get it. Thanks!

**Comment by atinux** [comment]
I would take a look at how we handle in the Sentry module as inspiration: https://github.com/nuxt-community/sentry-module/blob/master/lib/module.js#L60-L73

To test easily without having to write a module, you can directly use hooks directly in your `nuxt.config.js`, example:

```js
export default {
  hooks: {
    'render:errorMiddleware'(app) {
      app.use((err, req, res, next) => {
        // Catch the error with you error reported
        console.error(`An error occured while rendering ${req.url}\n`, err)
        // Let Nuxt show the error page
        // You can customize the HTML error page for production by creating app/views/error.html
        next(err)
      })
    }
  }
}
```

**Comment by Sletheren** [comment]
@claudiolcastro I' try to add newRelic to my project as well, an SSR based application, what I did is:
- Import `newRelic` in `store/nuxtServerInit`
- Registering `newrelic` as an external in webpack (same as you did)

But I'm having an error in the browser saying: **`ReferenceError: newrelic is not defined`**
Any idea? anyone?

**Comment by danielroe** [comment]
@Sletheren This might be relevant: https://github.com/nuxt/nuxt.js/issues/7960#issuecomment-832607715

**Comment by Sletheren** [comment]
@danielroe I tried the same setup, but I still get the same error `ReferenceError: newrelic is not defined` and noticed that the content of the script is `NREUM L(1)`
```
{
    hid: 'newrelic',
    innerHTML: '<!-- NREUM: (1) -->',
    type: 'text/javascript'
  }
```

**Comment by nachogarcia** [comment]
I didn't create a plugin because on further inspection it might not be neccesary to set the names on each request, though in my system I have the configuration that was quoted above.

Since the change in configuration from some versions ago, you can just try to import newrelic in top of nuxt.config.js.

And if you need further customization, use a hook like

```
hooks: {
    'render:route' (url, result, context) {
      try {
        newrelic.setTransactionName(url)
      } catch (error) {
        // Ignore
      }
    }
  }
```

Please let me know what worked for you, so we can create a final and updated answer to anyone trying to setup newrelic with nuxt, and make a plugin if deemed neccesary :)

**Reply by Sletheren** [reply]
@nachogarcia Thanks for your reply, I tried to do what you suggested, and here is a reproduction, https://github.com/Sletheren/nuxt-typescript-envs
When you run `yarn build` you will see that the building phase is failing due to "missing packages", sounds like its coming from `@newrelic/Winston-enricher`

**Comment by Christopher-Hayes** [comment]
If it helps, here's what we did: https://gist.github.com/Christopher-Hayes/60d03f2c8d274be31caf1658a84d3153

It pulls from different answers here. (Thank you! 🙏) Combined all changes we made into a single gist to make it a little easier for others.

**Comment by luksak** [comment]
Has anyone got a working solution for Nuxt 3?

**Reply by luksak** [reply]
I just published https://github.com/attributech/nuxt-newrelic that provides better integration between Nuxt and New Relic. This is my first Nuxt module and my first npm package. I'd be happy to receive some feedback on it 🫶
